cns/restserver/ipam.go

```diff
@@ -478,101 +627,230 @@ func (service *HTTPRestService) MarkExistingIPsAsPendingRelease(pendingIPIDs []s
+		case types.Assigned:
+			// This IP has already been assigned, if it is assigned to same pod add the IP to podIPInfo
+			if ipConfig.PodInfo.Key() == podInfo.Key() {
+				logger.Printf("[AssignDesiredIPConfigs]: IP Config [%+v] is already assigned to this Pod [%+v]", ipConfig, podInfo)
+
+				if err := service.populateIPConfigInfoUntransacted(ipConfig, &podIPInfo[numIPConfigsAssigned]); err != nil {
+					//nolint:goerr113 // return error
+					return []cns.PodIpInfo{}, fmt.Errorf("[AssignDesiredIPConfigs] Failed to assign IP %+v requested for pod %+v", ipConfig, podInfo)
 				}
-			default:
-				return podIpInfo, errors.Errorf("[AllocateDesiredIPConfig] Desired IP is not available %+v", ipConfig)
+				numIPConfigsAssigned++
+			} else {
+				//nolint:goerr113 // return error
+				return []cns.PodIpInfo{}, fmt.Errorf("[AssignDesiredIPConfigs] Desired IP is already assigned %+v, requested for pod %+v", ipConfig, podInfo)
 			}
-			err := service.populateIPConfigInfoUntransacted(ipConfig, &podIpInfo)
-			return podIpInfo, err
+		case types.Available, types.PendingProgramming:
+			// This race can happen during restart, where CNS state is lost and thus we have lost the NC programmed version
+			// As part of reconcile, we mark IPs as Assigned which are already assigned to Pods (listed from APIServer)
+			ipConfigsToAssign = append(ipConfigsToAssign, ipConfig)
+		default:
+			logger.Errorf("[AssignDesiredIPConfigs] Desired IP is not available %+v", ipConfig)
+			//nolint:goerr113 // return error
+			return podIPInfo, fmt.Errorf("IP not available")
+		}
+		// checks if found all of the desired IPs either as an available IP or already assigned to the pod
+		if len(ipConfigsToAssign)+numIPConfigsAssigned == numDesiredIPAddresses {
+			break
+		}
+	}
+
+	// if we did not find all of the desired IPs return an error
+	if len(ipConfigsToAssign)+numIPConfigsAssigned != numDesiredIPAddresses {
+		//nolint:goerr113 // return error
+		return podIPInfo, fmt.Errorf("not enough desired IPs found in pool")
+	}
+
+	failedToAssignIP := false
+	// assigns all IPs that were found as available to the pod
+	for i := range ipConfigsToAssign {
+		if err := service.assignIPConfig(ipConfigsToAssign[i], podInfo); err != nil {
+			logger.Errorf(err.Error())
+			failedToAssignIP = true
+			break
+		}
+		if err := service.populateIPConfigInfoUntransacted(ipConfigsToAssign[i], &podIPInfo[numIPConfigsAssigned]); err != nil {
+			logger.Errorf(err.Error())
+			failedToAssignIP = true
+			break
 		}
+		// adds to the newly assigned IP to the counter
+		numIPConfigsAssigned++
 	}
-	return podIpInfo, fmt.Errorf("Requested IP not found in pool")
+
+	// if we were able to get at least one IP but not all of the desired IPs
+	if failedToAssignIP {
+		logger.Printf("[AssignDesiredIPConfigs] Failed to retrieve all desired IPs. Releasing all IPs that were found")
+		for i := range ipConfigsToAssign {
+			_, err := service.unassignIPConfig(ipConfigsToAssign[i], podInfo)
```

Comment:
if we try to unassign IP which is not assigned, would that be an issue?

Reply:
This shouldn't cause any issues. It should just try to set the state of the IP to Available which shouldn't be an issue. For time's sake we could keep track of the ones successfully assigned and then only unassign those.

Reply:
yeah I looked at the code as well.  this should be fine